### PR TITLE
Prevent crash if resizing a newly formatted filesystem

### PR DESCRIPTION
**Pull request by shundhammer**
## Bugzilla

https://bugzilla.suse.com/show_bug.cgi?id=1124146

## Trello

https://trello.com/c/OUo0x3S3/

## Problem

In the partitioner, when the user changed an existing partition to be formatted and then immediately clicked on "resize", there would be a crash with a libstorage exception.

## Cause

The code tried to call `detect_size_info` on the filesystem, but that filesystem did not yet exist in the probed/system devicegraph.

## Fix

Now also checking if the planned filesystem is the same as the existing one or if it is new.

## Test

Manual test on my workstation with "yast2 disk" and a USB stick.
Changed the existing VFAT filesystem to ext4 and then hit _Resize_.

### CODE REVIEW

File: src/lib/y2partitioner/dialogs/blk_device_resize.rb [hunk 1]
```diff
@@ -94,6 +94,7 @@ def next_handler
 
       def detect_space_info
         return unless formatted? && controller.committed_device? && !swap?
+        return unless controller.committed_current_filesystem?
```
Comment (ancorgs) [inline src/lib/y2partitioner/dialogs/blk_device_resize.rb]: Why did you remove the `!swap?` part? As far as I understand the previous code, `Filesystem#detect_space_info` should not be called for swap filesystems. But that check is gone now.

File: src/lib/y2partitioner/dialogs/blk_device_resize.rb [hunk 2]
```diff
@@ -94,6 +94,7 @@ def next_handler
...
         @space_info = device.filesystem.detect_space_info
```
Comment (ancorgs) [inline src/lib/y2partitioner/dialogs/blk_device_resize.rb]: Moreover, to avoid similar crashes in the future, maybe we should do something like

```ruby
begin
  @space_info = device.filesystem.detect_space_info
rescue Storage::Exception
  log.error "Storage exception calling detect_space_info"
  # Should we alert the user in addition to registering the error?
  @space_info = nil
end
```

Uncatched libstorage-ng exceptions look uglier than exceptions thrown in Ruby. Moreover, displaying the widget without the "currently used" info is acceptable, so an error detecting the available space is actually not severe enough to interrupt the execution.

Maybe we could even inform the user that something weird happened checking that filesystem, in addition to just logging the error. Then is up to the user whether to try the resizing or not.

File: src/lib/y2partitioner/dialogs/blk_device_resize.rb (hunk 1 above)
Comment (shundhammer) [reply src/lib/y2partitioner/dialogs/blk_device_resize.rb]: I didn't. That's an extra line of code. Otherwise the entire expression would have been too long for Rubocop's liking.

File: src/lib/y2partitioner/dialogs/blk_device_resize.rb (hunk 2 above)
Comment (lslezak) [reply src/lib/y2partitioner/dialogs/blk_device_resize.rb]: If we cannot determine the space info then the FS is probably broken or in a bad shape. In that case resizing it might be very dangerous, it might fail or the data can be lost. In this case I'd really appreciate a warning message with "resizing it may fail or cause data loss" info.

File: src/lib/y2partitioner/dialogs/blk_device_resize.rb (hunk 1 above)
Comment (ancorgs) [reply src/lib/y2partitioner/dialogs/blk_device_resize.rb]: Right. I got confused because then the `formatted? && controller.committed_device?` part is absolutely redundant with the condition you added. Isn't it? This should be enough and more clear:

```ruby
def detect_space_info
  return unless controller.comitted_current_filesystem? && !swap?
  # the rest
end
```

Comment (shundhammer) [reply src/lib/y2partitioner/dialogs/blk_device_resize.rb]: I am not sure. I did the defensive approach. Can't there be other situations where only the old condition is true?

Comment (ancorgs) [reply src/lib/y2partitioner/dialogs/blk_device_resize.rb]: I'm pretty sure this is totally sure.

First, semantically:

- If `formatted?` is false, then there is no filesystem on the device. Thus `committed_current_filesystem?` MUST be false as well.
- If `controller.committed_device?` is false, then the device was not in the original devicegraph. Thus, its filesystem was not there either.

And now looking at the code:

- If `formatted?` is false, then `committed_current_filesystem?` is also false because that method starts with `return false unless device.formatted?`
- If `controller.committed_device?` is false, then `committed_current_filesystem?` is also false because it directly relies on `committed_filesystem` whose first line is `return nil unless committed_device?`.

Comment (shundhammer) [reply src/lib/y2partitioner/dialogs/blk_device_resize.rb]: OK then. Removing the extra conditions.

**Comment by shundhammer** [comment]
Warning popup:

detect-space-info-warning

**Comment by ancorgs** [comment]
Please merge https://github.com/shundhammer/yast-storage-ng/pull/1, everything else looks good.

**Comment by shundhammer** [comment]
Had to rebase from master.
